### ISSUE TITLE
[1.3-rc3] cordova-plugin-file Entry.toURL() no longer returns a filesystem URL on iOS

### ISSUE THREAD

**Issue by awwx** [meteor/meteor#6536]
On iOS (only), a file Entry `toURL()` method call is no longer returning a filesystem URL, but is now returning an `http://localhost:` URL.

On Android, `toURL()` continues to return a filesystem URL.

This appears to be [by design](http://guide.meteor.com/v1.3/1.3-migration.html#mobile):

> On iOS, core plugins like cordova-plugin-file and cordova-plugin-camera (using Camera.DestinationType.FILE_URI) will now automatically return these URLs instead of file:// URLs.

However, there are several problems here:

* Won't this break existing Cordova packages and plugins using `toURL()` and expecting the standard behavior? 

* According to the [W3C standard](https://dev.w3.org/2009/dap/file-system/file-dir-sys.html#widl-Entry-toURL-DOMString), `toURL()` should return a URL which "describes a location on disk".  

* Sometimes we need to get a filesystem URL from a file Entry.  For example, the filesystem URL is needed for the [upload method](https://www.npmjs.com/package/cordova-plugin-file-transfer#upload) in the Cordova file-transfer plugin.  If it's better for `toURL()` to return a `http://localhost:` URL, there should also be a new mechanism for obtaining the filesystem URL from a file Entry.

* The API is inconsistent between iOS and Android.  If Meteor provides a `http://localhost:` URL (whether through `toURL()` or by a different mechanism), it should do so consistently between platforms so that we don't need to write different code for the different platforms.

Unless there's some vital reason to munge `toURL()` on iOS, I'd recommend not breaking the standard.  After all, we do have `WebAppLocalServer.localFileSystemUrl()` to convert a `file://` URL when we need to.

## Details

In Meteor 1.2.1, calling `toURL()` on a file [Entry](https://dev.w3.org/2009/dap/file-system/file-dir-sys.html#the-entry-interface) returns a `file:///` URL on both iOS and Android.

iOS example: `file:///var/mobile/Containers/Data/Application/C9C2C6B3-4AE4-4FD7-9B50-09C677DB96A3/Library/NoCloud/`

Android example: `file:///data/data/com.id1gpk3qfl8wotm345bcc/files/`

In Meteor 1.3-rc3, `toURL()` is now returning an `http://localhost:` URL on iOS.

iOS example: `http://localhost:12728/local-filesystem/var/mobile/Containers/Data/Application/DB5588AA-E8C2-4336-B79D-79E739E7EC7C/Library/NoCloud/`

While Android is still returning the filesystem url.

Android example: `file:///data/data/com.idtsn2w384fq7x188ywqd/files/`

## Reproduction

```
$ meteor create url --release 1.2.1
$ cd url
$ meteor add-platform android
$ meteor add-platform ios
$ meteor add cordova:cordova-plugin-console@1.0.2
$ meteor add cordova:cordova-plugin-file@4.1.1
```

In client/url.js:

```
function reportError(err) {
  alert(err && err.stack || '' + err);
}

Meteor.startup(function () {
  console.log('---- client startup');
  Meteor.setTimeout(
    () => {
      resolveLocalFileSystemURL(
        cordova.file.dataDirectory,
        function (fileEntry) {
          console.log('resolve success')
          console.log(fileEntry.toURL());
        },
        reportError
      );
    },
    1000
  );
});
```

```
$ meteor run android-device
$ meteor run ios-device
```

To test with 1.3-rc.3:

```
meteor update --release 1.3-rc.3
```

**Comment by martijnwalraven** [comment]
Thanks for this well researched and well argued issue. You make a good point, so it might be worth changing this, even if it is at the last minute (we're really close to a release!).

I implemented that URL transformation mechanism on iOS because it is [also used](https://github.com/apache/cordova-plugins/blob/master/local-webserver/src/ios/CDVLocalWebServer.m#L167-L201) in the [local web server](https://github.com/apache/cordova-plugins/tree/master/local-webserver) developed as part of [`wkwebview-engine-localhost`](https://github.com/apache/cordova-plugins/tree/master/wkwebview-engine-localhost), which at least at one point was thought of as an official solution for running with WKWebView on iOS 8.

But you're right, it is of limited use for us because Android does not support a similar transformation mechanism. So you would have to run the resulting URL through `WebAppLocalServer.localFileSystemUrl()` anyway to be compatible with Android. And it might indeed break passing the URL to other plugins expecting a filesystem URL.

**Comment by awwx** [comment]
You're welcome!

Totally understandable of course that you want to get the release out on time.  On the other hand, the point of having release candidates is to find potential problems... :)

Since one of your goals is to have Meteor work well within the JavaScript ecosystem, it would be better to avoid making breaking changes to the public API of non-Meteor packages.  (Unless you really need to).

> And it might indeed break passing the URL to other plugins expecting a filesystem URL.

I haven't researched other plugins, but it did break my code :)

I could work around it naturally.  What I notice though is that the Cordova file API by itself is already a steep learning curve, where one needs to plow through three different (some discontinued!) W3C recommendations and then reconcile that with the Cordova specifics to figure out how to do simple file operations.  So it's already pretty painful.  And then if on top of that Meteor+Cordova works differently than plain Cordova...

What's the impact of not setting up the transformation for `toURL()`?  Is it a big change?

**Comment by martijnwalraven** [comment]
Not setting up the transformation is actually pretty easy, because Cordova exposes a nice hook specifically for this purpose. I think your arguments make sense, so I went ahead and removed the code (https://github.com/meteor/cordova-plugin-meteor-webapp/commit/d1f4f4c174d9c36745a9db2b7117e3ced40dbc74).

This should be in the next (last?) Meteor RC.

**Comment by awwx** [comment]
:grinning:

**Comment by benjamn** [comment]
Fixed in [`1.3-rc.5`](https://github.com/meteor/meteor/issues/6266#issuecomment-199632139), though I recommend updating to [`1.3-rc.6`](https://github.com/meteor/meteor/issues/6266#issuecomment-199903292) to avoid #6555.
